scripts/bootloader/provision.py

```diff
@@ -13,45 +13,92 @@
 from hashlib import sha256
 import os
 
-# Size of LCS storage in OTP in bytes
+# Size of LCS storage and implementation ID in OTP in bytes
 LCS_STATE_SIZE = 0x8
 IMPLEMENTATION_ID_SIZE = 0x20
 NUM_BYTES_PROVISIONED_ELSEWHERE = LCS_STATE_SIZE + IMPLEMENTATION_ID_SIZE
 
+# These variable names and values are copied from bl_storage.h and
+# should be kept in sync
+BL_MONOTONIC_COUNTERS_DESC_NSIB = 1
+BL_MONOTONIC_COUNTERS_DESC_MCUBOOT_ID0 = 2
+
+def generate_provision_intel_hex_file(provision_data, provision_address, output, max_size):
+    assert len(provision_data) <= max_size, """Provisioning data doesn't fit.
+Reduce the number of public keys or counter slots and try again."""
+
+    ih = IntelHex()
+    ih.frombytes(provision_data, offset=provision_address)
+    ih.write_hex_file(output)
+
+
+def add_hw_counters(provision_data, num_counter_slots_version, mcuboot_counters_slots):
+    num_counters = min(mcuboot_counters_slots, 1) + min(num_counter_slots_version, 1)
+
+    if num_counters == 0:
+        return provision_data
+
+    assert num_counter_slots_version % 2 == 0, "--num-counters-slots-version must be an even number"
+    assert mcuboot_counters_slots    % 2 == 0, "--mcuboot-counters-slots     must be an even number"
+
+    provision_data += struct.pack('H', 1) # Type "counter collection"
+    provision_data += struct.pack('H', num_counters) # Could be 0, 1, or 2
+
+    if num_counter_slots_version > 0:
+        provision_data += struct.pack('H', BL_MONOTONIC_COUNTERS_DESC_NSIB)
+        provision_data += struct.pack('H', num_counter_slots_version)
+        provision_data += bytes(2 * num_counter_slots_version * [0xFF])
+
+    if mcuboot_counters_slots > 0:
+        provision_data += struct.pack('H', BL_MONOTONIC_COUNTERS_DESC_MCUBOOT_ID0)
+        provision_data += struct.pack('H', mcuboot_counters_slots)
+        provision_data += bytes(2 * mcuboot_counters_slots * [0xFF])
+
+    return provision_data
+
+
+def generate_mcuboot_only_provision_hex_file(provision_address, output, max_size, mcuboot_counters_slots):
+    num_bytes_in_lcs = LCS_STATE_SIZE
```

Comment:
We could add an explanation in the function to describe the use case for this since it is not clear. Let's explain that because of the usage of the bl_storage library for the Mcuboot counters we need to place the counters in the offset that the bl_storage expects and these calculations here are just calculating this offset.

Reply:
fixed